### DISCUSSION TITLE
How do I get Quarto to work with R in Ubuntu?

### DISCUSSION THREAD

**Discussion by ThomasA** [quarto-dev/quarto-cli#2333] [Q&A] [answered]
I have installed Quarto in Ubuntu from <https://github.com/quarto-dev/quarto-cli/releases/download/v1.1.189/quarto-1.1.189-linux-amd64.deb>. I can run `quarto --help` afterwards.  
I try `quarto preview` on the template file <https://github.com/WarwickCIM/quarto-workshop/blob/main/cases/template/documentation_template.qmd>. This complains:

    Error: 'hooks_markdown' is not an exported object from 'namespace:knitr'

I have absolutely no experience with R, but I believe I have installed it. I have installed the packages "r-base" and "r-cran-rmarkdown" from the default Ubuntu repositories, using `apt`. `apt` reports that I also have "r-cran-knitr" installed (probably a consequence of "r-cran-rmarkdown").

I can enter R and run the command `library("knitr")` with no errors, so I assume knitr looks installed from R's perspective.

Does anyone have any idea how to get this to work?

**Comment by cderv** [comment, marked as answer]
Can you share which version of **knitr** has been installed using the default Ubuntu repo ? 

The function not found has been added in **knitr** 1.30 (https://github.com/yihui/knitr/blob/master/NEWS.md#changes-in-knitr-version-130)

I wonder if you have an older version somehow 🤔 

Instruction to install R on Ubuntu can be found on CRAN website: https://cran.r-project.org/bin/linux/ubuntu/fullREADME.html

Hope it helps

**Reply by ThomasA** [reply]
Oh, `apt` says:

    r-cran-knitr/focal,focal,now 1.28+dfsg-1 all [Installed]

so that could be the explanation (1.28).

**Reply by cderv** [reply]
Definitely, knitr 1.28 is too old. I am surprised we don't inform about that 🤔

**Reply by cderv** [reply]
Can you share with me the result of `quarto check` on your system please ?

**Reply by ThomasA** [reply]
Now I have added the "cran2deb4ubuntu" PPA on my system (https://launchpad.net/~c2d4u.team/+archive/ubuntu/c2d4u4.0+) and upgraded "r-base" and "r-cran-rmarkdown" afterwards. This seems to have changed something, but now R complains that Rmarkdown is not installed.

**Reply by ThomasA** [reply]
```
 quarto check

[✓] Checking Quarto installation......OK
      Version: 1.1.189
      Path: /opt/quarto/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.8.10
      Path: /usr/bin/python3
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 3.6.3
      Path: /usr/lib/R
      LibPaths:
        - /usr/local/lib/R/site-library
        - /usr/lib/R/site-library
        - /usr/lib/R/library
      rmarkdown: (None)

      The rmarkdown package is not available in this R installation.
      Install with install.packages("rmarkdown")
```

I am surprised that having the apt package "r-cran-rmarkdown" installed now does not seem visible to R.

**Reply by cderv** [reply]
It seems the PPA added is for R 4.0+ and quarto report 3.6.3. 
Probably all available and old in PATH ? 

I just remember a new project to install and manage R on all OS: https://github.com/r-lib/rig

Maybe that will be easier than installing through APT. You can also get more update package then by installing from R directly using `pak::pkg_install()` (https://pak.r-lib.org/)

Sorry to remember this only now.

**Reply by ThomasA** [reply]
I have also tried `install.packages("rmarkdown")` inside R now, but this complains with errors:
```
...
* installing *source* package ‘rmarkdown’ ...
** package ‘rmarkdown’ successfully unpacked and MD5 sums checked
** using staged installation
** R
** inst
** byte-compile and prepare package for lazy loading
Fejl i dyn.load(file, DLLpath = DLLpath, ...) : 
  kan ikke indlæse delt objekt '/usr/lib/R/site-library/rlang/libs/rlang.so':
  /usr/lib/R/site-library/rlang/libs/rlang.so: undefined symbol: R_ActiveBindingFunction
Kald: <Anonymous> ... namespaceImport -> loadNamespace -> library.dynam -> dyn.load
Kørsel stoppet
ERROR: lazy loading failed for package ‘rmarkdown’
* removing ‘/home/tari/R/x86_64-pc-linux-gnu-library/3.6/rmarkdown’

The downloaded source packages are in
	‘/tmp/Rtmpbz6q9p/downloaded_packages’
Advarselsbesked:
I install.packages("rmarkdown") :
  installation of package ‘rmarkdown’ had non-zero exit status
```

Sorry about the Danish, but it basically says:
```
Error in dyn.load(file, DLLpath = DLLpath, ...) : 
  cannot load shared object '/usr/lib/R/site-library/rlang/libs/rlang.so':
  /usr/lib/R/site-library/rlang/libs/rlang.so: undefined symbol: R_ActiveBindingFunction
Call: <Anonymous> ... namespaceImport -> loadNamespace -> library.dynam -> dyn.load
Runtime stopped
ERROR: lazy loading failed for package ‘rmarkdown’
* removing ‘/home/tari/R/x86_64-pc-linux-gnu-library/3.6/rmarkdown’

The downloaded source packages are in
	‘/tmp/Rtmpbz6q9p/downloaded_packages’
Warning message:
In install.packages("rmarkdown") :
  installation of package ‘rmarkdown’ had non-zero exit status
```

**Reply by ThomasA** [reply]
Wow, R sure is confusing...

**Reply by cderv** [reply]
Error above is about **rlang** package not correctly installed it seems. 

I would check your R version (3.6.3 or more recent), and then install from R the packages as you did. On Linux you may need system dependencies for some packages, so probably using **pak** will help. 

If **rig** + **pak** is an option, that could be the easiest (and not rely on Ubuntu packaging of R and R package)

**Reply by ThomasA** [reply]
My guess is that "r-base" and "r-cran-rmarkdown" may still be the packages from Ubuntu's default repos and that the CRAN PPA may have different names for some or all of the packages? I wonder how I can search specifically in packages from PPA. The (PPA) web page https://launchpad.net/~c2d4u.team/+archive/ubuntu/c2d4u4.0+ just lists 22486 package names with no apparent way to search through them in the browser...

**Reply by ThomasA** [reply]
I feel *really* uncomfortable having to resort to other package managers. I would really like to understand how to make `apt` get it right.

**Reply by ThomasA** [reply]
Do you happen to know what the R base package from the PPA should be called? I have installed the PPA and run `apt update` and all I get is:

    $ apt list r-base
    Listing... Done
    r-base/focal,focal 3.6.3-2 all

**Reply by ThomasA** [reply]
Maybe "r-cran-rsuite"?

**Reply by ThomasA** [reply]
Hmm, that cannot be installed due to loads of issues with dependencies.

**Reply by ThomasA** [reply]
Well, I will not be needing R anyway so I think I will just abandon it here and just try out Quarto with Python code only. I am used to Python and at least that works for me in Quarto.

**Reply by ThomasA** [reply]
Thanks for the suggestions, anyway.

**Reply by cderv** [reply]
Unfortunately, I don't know. Usually I am following the doc I shared (https://cran.r-project.org/bin/linux/ubuntu/fullREADME.html) and then managing my R package dependencies from R and not from `apt`. 

From my understanding, `r-cran-pkgname` should be the form to use (from https://cran.r-project.org/bin/linux/ubuntu/fullREADME.html#supported-packages) 
Be aware of that though
> The other r-cran-* packages are **updated with Ubuntu releases only. **

That is why I am realying on the official R package manager which is CRAN (or one of the mirror) and not a repackaging with less frequent update. 

In the PPA, by activating package details you can search. This is a result for **knitr**
https://launchpad.net/~c2d4u.team/+archive/ubuntu/c2d4u4.0+/+packages?field.name_filter=knitr&field.status_filter=published
I must says that I see `r-cran-knitr` is 1.28 but it seems there is a 1.40 version 🤔 

I really don't know about the other R packages than the `r-recommended` bundle. As I said I prefer to manage package from CRAN directly.

**Reply by ThomasA** [reply]
I guess installing packages from inside R would be OK, but having no experience with R, I don't know anything about how things are usually done in R. Would I need to install a 4-something version of R to get Quarto to work with knitr rmarkdown? R 4-something seems to be my major hurdle right now.

**Reply by ThomasA** [reply]
I cannot seem to install R according to https://cran.r-project.org/bin/linux/ubuntu/fullREADME.html#installation because the repository does not seem to be signed. At least I cannot find instructions on where to get the signature from. Instructions for adding apt repositories in Ubuntu usually come with a command for importing the repository's signature.

**Reply by eitsupi** [reply]
R on Ubuntu is really hard and I don't think it is possible for a beginner to build an environment.

I think the easy way for now is to use a package-installed Docker container or use r2u. (cc @eddelbuettel)
https://github.com/eddelbuettel/r2u

**Reply by eitsupi** [reply]
> I cannot seem to install R according to [cran.r-project.org/bin/linux/ubuntu/fullREADME.html#installation](https://cran.r-project.org/bin/linux/ubuntu/fullREADME.html#installation) because the repository does not seem to be signed. At least I cannot find instructions on where to get the signature from. Instructions for adding apt repositories in Ubuntu usually come with a command for importing the repository's signature.

I believe the following page linked from the R project describes it.
https://cran.r-project.org/bin/linux/ubuntu/

**Reply by ThomasA** [reply]
Thank you, that was exactly what I needed. I got confused by thinking https://cran.r-project.org/bin/linux/ubuntu/fullREADME.html#installation was where I needed to look, but the description at https://cran.r-project.org/bin/linux/ubuntu/ solves everything.

**Reply by cderv** [reply]
Good to know. Sorry to have pointed to the wrong page in the first place... 😞 Thanks @eitsupi !

**Comment by eddelbuettel** [comment]
Which Ubuntu version is this, and are you keeping CRAN up-to-date?

I am on 22.04, and have zero issues.  Also see https://dirk.eddelbuettel.com/blog/2022/01/31#035_apt_install_rstudio_quarto -- I now use quarto (and the two RStudio packages) off a 'truly personal PPA' as I got tired of dealing with outside of the (usually daily) `apt` upgrades.  That repo updates twice a week, and I happen to have gotten a new quarto package today.  (I cannot make the repo public as the files are larger than what GitHub allows so :disappointed: the actual repo is local on my box.)

**Reply by ThomasA** [reply]
I am still on 20.04 and do not know what keeping CRAN up to date means. Maybe it is finally time to upgrade to 22.04...

**Reply by ThomasA** [reply]
https://github.com/quarto-dev/quarto-cli/discussions/2333#discussioncomment-3587155 solved everything - wonderful :1st_place_medal:
